google/cloud/bigquery/client.py

```diff
@@ -1609,7 +1719,7 @@ def _get_query_results(
         )
         return _QueryResults.from_api_repr(resource)
 
-    def job_from_resource(self, resource):
+    def job_from_resource(self, resource: dict) -> job.UnknownJob:
```

Comment:
Can we do a union type here?

Alternatively, what do you think about introducing some `abc`-style module for all these different job classes? I know we have the [_AsyncJob](https://github.com/googleapis/python-bigquery/blob/6502a602337ae562652a20b20270949f2c9d5073/google/cloud/bigquery/job/base.py#L131) class, but that's private and also poorly named given that it's not usable in true `async` code.

Reply:
I kind of dislike "wide" union hints that clutter the code, thus introducing a base class for these type sounds reasonable. But perhaps in a separate PR, since that's an actual code refactoring, and not merely annotations that don't affect the runtime?